dark/reads.py

```diff
@@ -378,41 +414,62 @@ def propertyDetails(self):
-                seenStart = True
-            elif residue == '*':
-                # Stop codon. Yield an ORF, if it has non-zero length.
-                if ORFStart is not None and index - ORFStart > 0:
-                    # The ORF has non-zero length.
-                    yield AAReadORF(self, ORFStart, index, openLeft, False)
-                    ORFStart = None
-                # After a stop codon, we can no longer be open on the left
-                # and we have no longer seen a start codon.
-                openLeft = seenStart = False
-            else:
-                if (seenStart or openLeft) and ORFStart is None:
-                    ORFStart = index
 
-        # End of sequence. Yield the final ORF if there is one and it has
-        # non-zero length.
-        length = len(self.sequence)
-        if ((seenStart or openLeft) and ORFStart is not None and
-                length - ORFStart > 0):
-            yield AAReadORF(self, ORFStart, length, openLeft, True)
+        # Return open ORFs to the left and right and closed ORFs within the
+        # sequence.
+        if openORFs:
+            ORFStart = 0
+            inOpenORF = True  # open on the left
+            inORF = False
+
+            for index, residue in enumerate(self.sequence):
+                if residue == '*':
+                    if inOpenORF:
+                        if index:
+                            yield AAReadORF(self, ORFStart, index, True, False)
+                        inOpenORF = False
+                    if inORF:
+                        if ORFStart != index:
+                            yield AAReadORF(self, ORFStart, index,
+                                            False, False)
+                        inORF = False
+                if residue == 'M':
+                    if not inOpenORF and not inORF:
+                        ORFStart = index + 1
+                        inORF = True
+
+            # End of sequence. Yield the final ORF, open to the right, if there
+            # is one and it has non-zero length.
+            length = len(self.sequence)
+            if inOpenORF and length > 0:
+                yield AAReadORF(self, ORFStart, length, True, True)
+            elif inORF and ORFStart != index:
+                yield AAReadORF(self, ORFStart, length, False, True)
+
+        # Return only closed ORFs.
+        if not openORFs:
+            inORF = False
+
+            for index, residue in enumerate(self.sequence):
+                if residue == 'M':
+                    if not inORF:
+                        inORF = True
+                        ORFStart = index + 1
+                if residue == '*':
+                    if inORF:
+                        if ORFStart == index:
+                            inORF = False
+                        else:
+                            yield AAReadORF(self, ORFStart,
+                                            index, False, False)
+                            inORF = False
```

Comment:
You have this line in both halves of the `if`. Just use it once, outdented.

Reply:
done